File: packages/java-sdk/src/main/java/io/schlepengine/examples/ComprehensiveExample.java
package io.schlepengine.examples;

import io.schlepengine.SchlepClient;
import io.schlepengine.api.*;
import io.schlepengine.exceptions.ApiException;
import io.schlepengine.exceptions.ConfigurationException;
import io.schlepengine.types.*;

import java.io.File;
import java.io.IOException;
import java.util.*;

/**
 * Comprehensive example demonstrating all Schlep-engine SDK API clients.
 *
 * This example shows how to use:
 * - Data Processing API
 * - ML Pipeline API
 * - Analytics API
 * - Document Extraction API
 * - Data Quality API
 * - Storage API
 * - Monitoring API
 * - Users API
 * - Admin API
 */
public class ComprehensiveExample {

    public static void main(String[] args) {
        try {
            // Initialize the client
            SchlepClient client = new SchlepClient("your-api-key-here");

            System.out.println("=== Schlep-engine Java SDK - Comprehensive Example ===\n");

            // 1. Data Processing API
            demonstrateDataProcessing(client);

            // 2. ML Pipeline API
            demonstrateMLPipeline(client);

            // 3. Analytics API
            demonstrateAnalytics(client);

            // 4. Document Extraction API
            demonstrateDocumentExtraction(client);

            // 5. Data Quality API
            demonstrateDataQuality(client);

            // 6. Storage API
            demonstrateStorage(client);

            // 7. Monitoring API
            demonstrateMonitoring(client);

            // 8. Users API
            demonstrateUsers(client);

            // 9. Admin API (requires admin privileges)
            demonstrateAdmin(client);

            // Clean up
            client.close();
            System.out.println("\n=== All examples completed successfully ===");

        } catch (ConfigurationException e) {
            System.err.println("Configuration error: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void demonstrateDataProcessing(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Data Processing API ---");

        DataProcessingClient dataClient = client.data();

        // Process a file
        File dataFile = new File("sample-data.csv");
        if (dataFile.exists()) {
            DataProcessingResult result = dataClient.processFile(
                dataFile,
                DataFormat.CSV,
                DataFormat.JSON
            );
            System.out.println("Processing job started: " + result.getJobId());

            // Check job status
            JobInfo jobStatus = dataClient.getJobStatus(result.getJobId());
            System.out.println("Job status: " + jobStatus.getStatus());
        }

        // List processing jobs
        List<JobInfo> jobs = dataClient.listJobs(1, 10);
        System.out.println("Total jobs found: " + jobs.size());

        System.out.println();
    }

    private static void demonstrateMLPipeline(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- ML Pipeline API ---");

        MLPipelineClient mlClient = client.ml();

        // Create a pipeline
        Map<String, Object> pipelineConfig = new HashMap<>();
        pipelineConfig.put("name", "Customer Classification Pipeline");
        pipelineConfig.put("task_type", MLTaskType.CLASSIFICATION.getValue());
        pipelineConfig.put("model_type", "random_forest");
        pipelineConfig.put("target_column", "customer_type");

        Map<String, Object> pipeline = mlClient.createPipeline(pipelineConfig);
        System.out.println("Pipeline created: " + pipeline.get("pipeline_id"));

        // List pipelines
        List<Map<String, Object>> pipelines = mlClient.listPipelines(1, 10);
        System.out.println("Total pipelines: " + pipelines.size());

        // Train a pipeline (if you have a pipeline ID)
        // String pipelineId = (String) pipeline.get("pipeline_id");
        // TrainingJob job = mlClient.trainPipeline(pipelineId, null, null);
        // System.out.println("Training job started: " + job.getJobId());

        // List models
        List<ModelInfo> models = mlClient.listModels(1, 10);
        System.out.println("Total models: " + models.size());

        System.out.println();
    }

    private static void demonstrateAnalytics(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Analytics API ---");

        AnalyticsClient analyticsClient = client.analytics();

        // Get available datasets
        List<Map<String, Object>> datasets = analyticsClient.getDatasets();
        System.out.println("Available datasets: " + datasets.size());

        // Run analytics query
        Map<String, Object> query = new HashMap<>();
        query.put("dataset", "sales_data");
        query.put("metrics", Arrays.asList("sum", "average", "count"));
        query.put("dimensions", Arrays.asList("region", "product_category"));
        query.put("filters", new HashMap<String, Object>() {{
            put("date_range", "last_30_days");
        }});

        // Map<String, Object> results = analyticsClient.query(query);
        // System.out.println("Query results: " + results);

        System.out.println();
    }

    private static void demonstrateDocumentExtraction(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Document Extraction API ---");

        DocumentClient docClient = client.document();

        // Extract text from PDF
        File pdfFile = new File("sample-document.pdf");
        if (pdfFile.exists()) {
            Map<String, Object> textResult = docClient.extractText(pdfFile, true, false);
            System.out.println("Text extraction completed");

            // Extract tables
            Map<String, Object> tablesResult = docClient.extractTables(pdfFile);
            System.out.println("Table extraction completed");
        } else {
            System.out.println("Sample document not found, skipping extraction");
        }

        System.out.println();
    }

    private static void demonstrateDataQuality(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Data Quality API ---");

        QualityClient qualityClient = client.quality();

        // Assess data quality
        List<String> checks = Arrays.asList("completeness", "validity", "uniqueness");
        // Map<String, Object> qualityReport = qualityClient.assessQuality(
        //     "s3://my-bucket/data.csv",
        //     checks
        // );
        // System.out.println("Quality report: " + qualityReport.get("report_id"));

        System.out.println();
    }

    private static void demonstrateStorage(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Storage API ---");

        StorageClient storageClient = client.storage();

        // Upload a file
        File file = new File("data.csv");
        if (file.exists()) {
            FileUpload upload = storageClient.uploadFile(file, "datasets");
            System.out.println("File uploaded: " + upload.getFileId());
        }

        // List files
        List<FileUpload> files = storageClient.listFiles(1, 20);
        System.out.println("Total files in storage: " + files.size());

        System.out.println();
    }

    private static void demonstrateMonitoring(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Monitoring API ---");

        MonitoringClient monitoringClient = client.monitoring();

        // Get system health
        Map<String, Object> health = monitoringClient.getSystemHealth();
        System.out.println("System status: " + health.get("status"));

        // Get metrics
        List<String> metricNames = Arrays.asList("cpu_usage", "memory_usage", "request_rate");
        Map<String, Object> metrics = monitoringClient.getMetrics(metricNames, null);
        System.out.println("Metrics retrieved: " + metrics.keySet().size());

        System.out.println();
    }

    private static void demonstrateUsers(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Users API ---");

        UsersClient usersClient = client.users();

        // Get user profile
        Map<String, Object> profile = usersClient.getProfile();
        System.out.println("User ID: " + profile.get("user_id"));
        System.out.println("User email: " + profile.get("email"));

        // Update profile
        Map<String, Object> updates = new HashMap<>();
        updates.put("name", "Updated Name");
        // Map<String, Object> updatedProfile = usersClient.updateProfile(updates);

        System.out.println();
    }

    private static void demonstrateAdmin(SchlepClient client)
            throws ApiException, IOException {
        System.out.println("--- Admin API (requires admin privileges) ---");

        AdminClient adminClient = client.admin();

        try {
            // Get system statistics
            Map<String, Object> stats = adminClient.getSystemStats();
            System.out.println("Total users: " + stats.get("total_users"));
            System.out.println("Total jobs: " + stats.get("total_jobs"));

            // List users
            List<Map<String, Object>> users = adminClient.listUsers(1, 10);
            System.out.println("User count: " + users.size());
        } catch (ApiException e) {
            if (e.getStatusCode() == 403) {
                System.out.println("Admin access required (403 Forbidden)");
            } else {
                throw e;
            }
        }

        System.out.println();
    }
}

File: labs/packages/csharp-sdk/Examples/ComprehensiveExample.cs
using SchlepEngine;
using SchlepEngine.Types;

namespace SchlepEngine.Examples;

/// <summary>
/// Comprehensive example demonstrating all Schlep-engine SDK API modules.
/// </summary>
/// <remarks>
/// This example showcases the complete API surface of the C# SDK including:
/// - Data Processing
/// - ML Pipeline
/// - Analytics
/// - Document Extraction
/// - Data Quality
/// - Storage
/// - Monitoring
/// - Users
/// - Admin
/// </remarks>
public class ComprehensiveExample
{
    public static async Task Main(string[] args)
    {
        // Initialize the client with your API key
        var client = new SchlepClient("your-api-key");

        try
        {
            // =============================================================
            // 1. DATA PROCESSING API
            // =============================================================
            Console.WriteLine("=== Data Processing API ===");

            // Process a file
            var fileData = await File.ReadAllBytesAsync("data.csv");
            var processingJob = await client.Data.ProcessFileAsync(fileData, "csv");
            Console.WriteLine($"Processing job created: {processingJob.JobId}");

            // Check job status
            var jobStatus = await client.Data.GetJobAsync(processingJob.JobId);
            Console.WriteLine($"Job status: {jobStatus.Status}");

            // List all processing jobs
            var jobs = await client.Data.ListJobsAsync(new ListParams
            {
                Page = 1,
                PageSize = 10,
                Status = "completed"
            });
            Console.WriteLine($"Found {jobs.Count} jobs");

            // Transform data
            var transformations = new
            {
                operations = new[]
                {
                    new { type = "filter", column = "age", condition = "> 18" },
                    new { type = "rename", from = "old_name", to = "new_name" }
                }
            };
            var transformResult = await client.Data.TransformDataAsync(processingJob.JobId, transformations);
            Console.WriteLine($"Transformation status: {transformResult.Status}");

            // Validate schema
            var schema = new
            {
                fields = new[]
                {
                    new { name = "id", type = "integer", required = true },
                    new { name = "name", type = "string", required = true }
                }
            };
            var validationResult = await client.Data.ValidateSchemaAsync(processingJob.JobId, schema);
            Console.WriteLine($"Validation result: {(validationResult.Valid ? "Valid" : "Invalid")}");

            // =============================================================
            // 2. ML PIPELINE API
            // =============================================================
            Console.WriteLine("\n=== ML Pipeline API ===");

            // Create a pipeline
            var pipelineConfig = new
            {
                name = "Customer Churn Prediction",
                model_type = "classification",
                task_type = "supervised",
                target_column = "churned",
                features = new[] { "age", "tenure", "monthly_charges" }
            };
            var pipeline = await client.ML.CreatePipelineAsync(pipelineConfig);
            Console.WriteLine($"Pipeline created: {pipeline.PipelineId}");

            // Train the pipeline
            var trainingConfig = new
            {
                test_size = 0.2,
                random_state = 42,
                hyperparameters = new
                {
                    max_depth = 10,
                    n_estimators = 100
                }
            };
            var trainingJob = await client.ML.TrainPipelineAsync(pipeline.PipelineId, trainingConfig);
            Console.WriteLine($"Training job started: {trainingJob.JobId}");

            // Check training status
            var trainingStatus = await client.ML.GetTrainingJobAsync(trainingJob.JobId);
            Console.WriteLine($"Training status: {trainingStatus.Status}");
            if (trainingStatus.Metrics != null)
            {
                Console.WriteLine($"Metrics: {string.Join(", ", trainingStatus.Metrics.Select(m => $"{m.Key}={m.Value}"))}");
            }

            // List all pipelines
            var pipelines = await client.ML.ListPipelinesAsync(new ListParams { Page = 1, PageSize = 10 });
            Console.WriteLine($"Found {pipelines.Count} pipelines");

            // Deploy model (if training completed)
            if (trainingStatus.ModelId != null)
            {
                var deployment = await client.ML.DeployModelAsync(trainingStatus.ModelId);
                Console.WriteLine($"Model deployed at: {deployment.EndpointUrl}");

                // Make predictions
                var predictionData = new
                {
                    age = 35,
                    tenure = 24,
                    monthly_charges = 79.99
                };
                var prediction = await client.ML.PredictAsync(deployment.EndpointUrl, predictionData);
                Console.WriteLine($"Predictions: {prediction.Predictions.Count}");

                // Get model information
                var modelInfo = await client.ML.GetModelAsync(trainingStatus.ModelId);
                Console.WriteLine($"Model: {modelInfo.Name} v{modelInfo.Version}");

                // List all models
                var models = await client.ML.ListModelsAsync(new ListParams { Page = 1, PageSize = 10 });
                Console.WriteLine($"Found {models.Count} models");
            }

            // =============================================================
            // 3. ANALYTICS API
            // =============================================================
            Console.WriteLine("\n=== Analytics API ===");

            // Execute analytics query
            var analyticsQuery = new
            {
                dataset = "sales",
                metrics = new[] { "revenue", "count" },
                dimensions = new[] { "region", "product" },
                filters = new[]
                {
                    new { field = "date", operator = ">=", value = "2024-01-01" }
                },
                aggregations = new[]
                {
                    new { field = "revenue", function = "sum" }
                }
            };
            var queryResult = await client.Analytics.QueryAsync(analyticsQuery);
            Console.WriteLine($"Query returned {queryResult.RowCount} rows in {queryResult.ExecutionTimeMs}ms");

            // Generate a report
            var reportConfig = new
            {
                name = "Monthly Sales Report",
                dataset = "sales",
                template = "monthly_summary",
                format = "pdf"
            };
            var report = await client.Analytics.GenerateReportAsync(reportConfig);
            Console.WriteLine($"Report generated: {report.ReportId}");

            // Get available datasets
            var datasets = await client.Analytics.GetDatasetsAsync();
            Console.WriteLine($"Available datasets: {datasets.Count}");

            // Get dataset schema
            if (datasets.Count > 0)
            {
                var schemaInfo = await client.Analytics.GetSchemaAsync(datasets[0].Name);
                Console.WriteLine($"Schema for {schemaInfo.Dataset}: {schemaInfo.Fields.Count} fields");
            }

            // =============================================================
            // 4. DOCUMENT EXTRACTION API
            // =============================================================
            Console.WriteLine("\n=== Document Extraction API ===");

            // Extract text from document
            var documentData = await File.ReadAllBytesAsync("document.pdf");
            var extraction = await client.Document.ExtractTextAsync(
                documentData,
                "document.pdf",
                extractTables: true,
                extractImages: false
            );
            Console.WriteLine($"Extracted text: {extraction.Text.Length} characters");
            Console.WriteLine($"Pages processed: {extraction.PageCount}");

            // Extract tables
            var tableExtraction = await client.Document.ExtractTablesAsync(documentData, "document.pdf");
            Console.WriteLine($"Extracted {tableExtraction.TableCount} tables");

            // Extract images
            var imageExtraction = await client.Document.ExtractImagesAsync(documentData, "document.pdf");
            Console.WriteLine($"Extracted {imageExtraction.ImageCount} images");

            // Perform OCR
            var imageData = await File.ReadAllBytesAsync("scanned.jpg");
            var ocrResult = await client.Document.PerformOCRAsync(imageData, "scanned.jpg", "en");
            Console.WriteLine($"OCR text: {ocrResult.Text.Length} characters");
            Console.WriteLine($"Confidence: {ocrResult.Confidence:P2}");

            // =============================================================
            // 5. DATA QUALITY API
            // =============================================================
            Console.WriteLine("\n=== Data Quality API ===");

            // Assess data quality
            var qualityAssessment = await client.Quality.AssessQualityAsync(
                "dataset-path",
                new[] { "completeness", "accuracy", "consistency" }
            );
            Console.WriteLine($"Quality score: {qualityAssessment.OverallScore}/100");
            Console.WriteLine($"Issues found: {qualityAssessment.Issues.Count}");

            // Create quality rule
            var qualityRule = new
            {
                name = "Email Validation",
                type = "pattern",
                config = new
                {
                    field = "email",
                    pattern = @"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
                }
            };
            var rule = await client.Quality.CreateRuleAsync(qualityRule);
            Console.WriteLine($"Quality rule created: {rule.RuleId}");

            // List quality rules
            var rules = await client.Quality.ListRulesAsync(new ListParams { Page = 1, PageSize = 10 });
            Console.WriteLine($"Found {rules.Count} quality rules");

            // Validate data against rules
            var validationResult2 = await client.Quality.ValidateAsync(
                "dataset-path",
                new[] { rule.RuleId }
            );
            Console.WriteLine($"Validation: {(validationResult2.Passed ? "Passed" : "Failed")}");
            Console.WriteLine($"Rules evaluated: {validationResult2.RulesEvaluated}");

            // =============================================================
            // 6. STORAGE API
            // =============================================================
            Console.WriteLine("\n=== Storage API ===");

            // Upload file
            var uploadData = await File.ReadAllBytesAsync("data.csv");
            var uploadResult = await client.Storage.UploadFileAsync(
                uploadData,
                "data.csv",
                folder: "datasets"
            );
            Console.WriteLine($"File uploaded: {uploadResult.FileId}");
            Console.WriteLine($"URL: {uploadResult.Url}");

            // List files
            var files = await client.Storage.ListFilesAsync(
                folder: "datasets",
                parameters: new ListParams { Page = 1, PageSize = 10 }
            );
            Console.WriteLine($"Found {files.Count} files");

            // Get file metadata
            var fileMetadata = await client.Storage.GetFileAsync(uploadResult.FileId);
            Console.WriteLine($"File: {fileMetadata.Filename} ({fileMetadata.Size} bytes)");

            // Download file
            var downloadedData = await client.Storage.DownloadFileAsync(uploadResult.FileId);
            Console.WriteLine($"Downloaded {downloadedData.Length} bytes");

            // =============================================================
            // 7. MONITORING API
            // =============================================================
            Console.WriteLine("\n=== Monitoring API ===");

            // Check system health
            var health = await client.Monitoring.GetSystemHealthAsync();
            Console.WriteLine($"System status: {health.Status}");
            if (health.Components != null)
            {
                foreach (var component in health.Components)
                {
                    Console.WriteLine($"  {component.Key}: {component.Value.Status}");
                }
            }

            // Get metrics
            var metrics = await client.Monitoring.GetMetricsAsync(
                metricNames: new[] { "cpu_usage", "memory_usage", "request_rate" },
                timeRange: new TimeRange
                {
                    Start = DateTime.UtcNow.AddHours(-1).ToString("o"),
                    End = DateTime.UtcNow.ToString("o")
                }
            );
            Console.WriteLine($"Retrieved {metrics.Metrics.Count} metrics");

            // Create alert
            var alertConfig = new
            {
                name = "High CPU Usage",
                severity = "high",
                condition = new
                {
                    metric = "cpu_usage",
                    threshold = 80,
                    operator = ">"
                }
            };
            var alert = await client.Monitoring.CreateAlertAsync(alertConfig);
            Console.WriteLine($"Alert created: {alert.AlertId}");

            // Get alerts
            var alerts = await client.Monitoring.GetAlertsAsync(new ListParams { Page = 1, PageSize = 10 });
            Console.WriteLine($"Found {alerts.Count} alerts");

            // =============================================================
            // 8. USERS API
            // =============================================================
            Console.WriteLine("\n=== Users API ===");

            // Get user profile
            var profile = await client.Users.GetProfileAsync();
            Console.WriteLine($"User: {profile.Email}");
            Console.WriteLine($"Role: {profile.Role}");
            Console.WriteLine($"Status: {profile.Status}");

            // Update profile
            var updates = new { name = "John Doe", organization = "Acme Corp" };
            var updatedProfile = await client.Users.UpdateProfileAsync(updates);
            Console.WriteLine($"Profile updated: {updatedProfile.Name}");

            // List API keys
            var apiKeys = await client.Users.ListApiKeysAsync();
            Console.WriteLine($"API keys: {apiKeys.Count}");

            // Create new API key
            var newKey = await client.Users.CreateApiKeyAsync(
                "Production Key",
                permissions: new[] { "read", "write" }
            );
            Console.WriteLine($"New API key created: {newKey.KeyId}");

            // =============================================================
            // 9. ADMIN API (requires admin privileges)
            // =============================================================
            Console.WriteLine("\n=== Admin API ===");

            // Get system statistics
            var stats = await client.Admin.GetSystemStatsAsync();
            Console.WriteLine($"Total users: {stats.TotalUsers}");
            Console.WriteLine($"Active users: {stats.ActiveUsers}");
            Console.WriteLine($"Total jobs: {stats.TotalJobs}");
            Console.WriteLine($"Storage used: {stats.StorageUsed / (1024 * 1024 * 1024)}GB");

            // List all users (admin only)
            var allUsers = await client.Admin.ListUsersAsync(new ListParams { Page = 1, PageSize = 20 });
            Console.WriteLine($"Total users in system: {allUsers.Count}");

            // Get specific user
            if (allUsers.Count > 0)
            {
                var user = await client.Admin.GetUserAsync(allUsers[0].UserId);
                Console.WriteLine($"User details: {user.Email} - {user.Status}");
            }

            Console.WriteLine("\n=== All API modules demonstrated successfully! ===");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            if (ex.InnerException != null)
            {
                Console.WriteLine($"Inner exception: {ex.InnerException.Message}");
            }
        }
        finally
        {
            // Clean up
            client.Dispose();
        }
    }
}
